placeSpecificContent/hq/courtyard/controlPanel/controlPanel_server.js

```diff
@@ -0,0 +1,403 @@
+    var EPSILON_FOR_COORDINATE_VALUES = 0.005;
+
+    var _this;
+
+    var injector;
+
+    var preloadIntervalCount = 0;
+    var dataRetrievalTimeout;
+    var ready = false;
+
+    var roundtable;
+    var courtyard;
+
+    var tableHeight;
+
+    var courtyardLoweredPosition;
+    var roundtableLoweredPosition;
+
+    var roundtableYVelocity;
+    var courtyardYVelocity;
+
+    var roundtableInPlace;
+    var courtyardInPlace;
+
+    var checkInterval;
+
+    var soundPosition;
+    
+
+    var Button = function() {
+        _this = this;
+    };
+
+    Button.prototype = {
+        remotelyCallable: ['pressButton'],
+
+        /* Save a reference to this and call another function to begin trying to read entity data */
+        preload: function(entityID) {
+            _this.entityID = entityID;
+            _this.setDataRequestTimeout();
+        },
+
+        /* Try to get entity data. If all data is retrieved or more than 10 attempts are made, stop trying. */
+        preloadSetupAttempt: function() {
+            preloadIntervalCount++;
+            try {
+                _this.getEntityData();
+            } catch (err) {
+                print("Attempt ", preloadIntervalCount, " to get entity data failed.");
+            }
+            if (JSON.stringify(BUTTON_TO_ENTITY_POSITIONING_MAP) !== "{}" && roundtable && courtyard) {
+                ready = true;
+            } else if (preloadIntervalCount < PRELOAD_DATA_RETRIEVAL_MAX_ATTEMPTS) {
+                _this.setDataRequestTimeout();
+            }
+        },
+
+        /* Set a timeout for 500ms after which we will attempt to read entity data */
+        setDataRequestTimeout: function() {
+            dataRetrievalTimeout = Script.setTimeout(function() {
+                dataRetrievalTimeout = null;
+                _this.preloadSetupAttempt();
+            }, DATA_RETRIEVAL_TIMEOUT_MS);
+        },
+
+        /* Use parenting relationships to get IDs of required entities and userData of this entity. Use the user data 
+        and IDs of buttons to create a map for where each entity should be at each state. Save the lowered position 
+        of each entity and the height of the table. If unable to get all necessary data, erase everything so we are 
+        not left with partial data that will cause errors. */
+        getEntityData: function() {
+            if (JSON.stringify(BUTTON_TO_ENTITY_POSITIONING_MAP) === "{}") {
+
+                var controlPanelProperties = Entities.getEntityProperties(_this.entityID, 'userData');
+                var controlPanelUserData = JSON.parse(controlPanelProperties.userData);
+
+                Entities.getChildrenIDs(_this.entityID).forEach(function(controlPanelChildID) {
+
+                    var name = Entities.getEntityProperties(controlPanelChildID, 'name').name;
+                    if (name.indexOf("Control Panel Button") > -1) {
+
+                        var stateName = name.substr(INDEX_WHERE_STATE_NAME_BEGINS).toLowerCase();
```

Comment:
Why can't you use `indexOf()` instead of relying on the entity name matching 100%?

Reply:
I can. I thought it would be better to be more exact in case anything else gets parented to the control panel at any point to ensure there are no mix-ups.

Reply:
ahhh, that makes sense. OK. i'm fine with this :)